### PR TITLE
added to genesis autonity contract section

### PR DESCRIPTION
**Pull request by b00ris**
Closes https://github.com/clearmatics/autonity/issues/186

Autonity contract section integration should be made after https://github.com/clearmatics/autonity/issues/169 in https://github.com/clearmatics/autonity/issues/191

### CODE REVIEW

**Comment by yazzaoui** [comment]
I was thinking that it was required to have "autonityContract" mandatory because we need to remove "validators" but then I noticed that we didn't have the distinction between validators and stake-holder in the srs which is clearly confusing .. That's why prototyping is important.

@aesmonty I think we need to introduce a new user class for "validators" to make it clear that stake-holder might not validate.

**Comment by aesmonty** [comment]
@yazzaoui Makes sense for me. Then, the difference would be:

- `member`: Can own stake token.
- `validator`: Can own stake token and can be selected to be part of the committee.

In A0.5 there is no committee so all `validators` are effectively part of the committee and participate in consensus voting.

File: core/autonity_contract_genesis.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/clearmatics/autonity/common"
+	"github.com/clearmatics/autonity/p2p/enode"
+	"reflect"
+)
+
+const (
+	//participant: Authorized to operate a full node, is able to join the network, is not authorized to own stake.
+	UserParticipant UserType = "participant"
+	//member: Authorized to operate a full node, is able to join the network, authorized to own stake.
+	UserMember UserType = "member"
```
Comment (yazzaoui) [inline core/autonity_contract_genesis.go]: From https://github.com/clearmatics/doc-srs-008/issues/2 we have to rename "member" to "stakeholder"

File: core/autonity_contract_genesis.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+// Autonity contract config. It'is used for deployment.
+type AutonityContract struct {
+	// Address of the validator who deploys contract stored in bytecode
+	Deployer common.Address `json:"deployer" toml:",omitempty"`
+	// Bytecode of validators contract // would like this type to be []byte but the unmarshalling is not working
+	Bytecode string `json:"bytecode" toml:",omitempty"`
+	// Json ABI of the contract
+	ABI                string         `json:"abi "toml:",omitempty"`
+	MinGasPrice        uint64         `json:"minGasPrice" toml:",omitempty"`
+	GovernanceOperator common.Address `json:"governanceOperator" toml:",omitempty"`
```
Comment (yazzaoui) [inline core/autonity_contract_genesis.go]: https://github.com/clearmatics/doc-srs-008/issues/2 : need to be renamed to "Operator"

File: core/autonity_contract_genesis.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+	if u.Type == UserParticipant && u.Stake > 0 {
+		return errors.New("user.stake must be nil or equal to 0 for users of type participant")
+	}
+
+	if u.Type == UserValidator && len(u.Enode) == 0 {
+		return errors.New("if user.type is validator then user.enode must be defined")
+	}
+	if len(u.Enode) > 0 {
+		n, err := enode.ParseV4(u.Enode)
```
Comment (yazzaoui) [inline core/autonity_contract_genesis.go]: I'm not sure if `enode.ParseV4` is compatible with enode using domain names, maybe @JekaMas can confirm?

Comment (yazzaoui) [reply core/autonity_contract_genesis.go]: probably it should be `ParseV4WithResolve`

Comment (JekaMas) [reply core/autonity_contract_genesis.go]: Yeap, it's better to use `ParseV4WithResolve` and write a few tests.

**Comment by b00ris** [comment]
@yazzaoui @JekaMas fixed

**Pull request merged**